libnd4j/include/ops/declarable/helpers/cuda/qr.cu

```diff
@@ -0,0 +1,176 @@
+
+    template <typename T>
+    NDArray vmul(LaunchContext* context, NDArray const& v, int n)
+    {
+        NDArray res('c', {n,n}, v.dataType(), context); // x = matrix_new(n, n);
+
+        auto stream = context->getCudaStream();
+        vmulKernel<T><<<128, 128, 128, *stream>>>(res.dataBuffer()->specialAsT<T>(), res.specialShapeInfo(),
+                reinterpret_cast<T const*>(v.getSpecialBuffer()), v.getSpecialShapeInfo(), n);
+        return res;
+    }
+
+    template <typename T>
+    static bool diagonalIsPositive(NDArray* matrix, Nd4jLong k) {
+        T hVal;
+        Nd4jLong pos[] = {k, k};
+        auto shift = shape::getOffset(matrix->shapeInfo(), pos);
+        cudaMemcpy(&hVal, matrix->specialBuffer(), sizeof(T), cudaMemcpyDeviceToHost);
+        return hVal > T(0.f);
+    }
+
+    template <typename T>
+    void qrSingle(LaunchContext* context, NDArray* matrix, NDArray* Q, NDArray* R, bool const fullMatricies) {
+        Nd4jLong M = matrix->sizeAt(-2);
+        Nd4jLong N = matrix->sizeAt(-1);
+        auto resQ = fullMatricies?Q->ulike():NDArrayFactory::create<T>(matrix->ordering(), {M,M}, Q->getContext());
+        auto resR = fullMatricies?R->ulike():matrix->ulike();
+        NDArray q[M];
+        NDArray z = *matrix;
+        for (auto k = 0; k < N && k < M - 1; k++) { // loop for columns, but not further then row number
+            NDArray e('c', {M}, DataTypeUtils::fromT<T>()); // two internal buffers and scalar for squared norm
+            z = matrixMinor<T>(context, z, k); // minor computing for current column with given matrix z (initally is a input matrix)
+            z.tickWriteDevice();
+            auto currentColumn = z({0, 0, k, k + 1}); // retrieve k column from z to x buffer
+            auto norm = currentColumn.reduceAlongDimension(reduce::Norm2, {0});
+            if (diagonalIsPositive<T>(matrix, k)) //matrix->t<T>(k,k) > T(0.f)) // negate on positive matrix diagonal element
+                norm.applyTransform(transform::Neg, norm); // *= -1.f;//-norm.t<T>(0);
+
+            e.p(k, norm); // e - is filled by 0 vector except diagonal element (filled by 1)
+            e += currentColumn; // e[i] = x[i] + a * e[i] for each i from 0 to n - 1
+            auto normE = e.reduceAlongDimension(reduce::Norm2, {0});
+            e /= normE;
+            q[k] = vmul<T>(context, e, M);
+            auto qQ = z.ulike();
+            MmulHelper::matmul(&q[k], &z, &qQ, false, false);
+            z = std::move(qQ);
+        }
+        resQ.assign(q[0]); //
+//        MmulHelper::matmul(&q[0], matrix, &resR, false, false);
+        for (int i = 1; i < N && i < M - 1; i++) {
+            auto tempResQ = resQ;
+            MmulHelper::matmul(&q[i], &resQ, &tempResQ, false, false);
+            resQ = std::move(tempResQ);
+        }
+        MmulHelper::matmul(&resQ, matrix, &resR, false, false);
+        // resR *= -1.f;
+        resQ.transposei();
+        resQ.tickWriteDevice();
+        resR.tickWriteDevice();
```

Comment:
Why you're manually ticking on device side here? resR should be ticked within matmul. and resQ should be updated in matmul on line 130 as well